contracts/VaultV3.vy

```diff
@@ -1187,24 +1187,59 @@ def process_report(strategy: address) -> (uint256, uint256):
 
 @external
 @nonreentrant("lock")
-def sweep(token: address) -> (uint256):
-    """
-    @notice Sweep the token from airdop or sent by mistake.
-    @param token The token to sweep.
-    @return The amount of dust swept.
-    """
-    self._enforce_role(msg.sender, Roles.SWEEPER)
-    assert token != self, "can't sweep self"
-    assert self.strategies[token].activation == 0, "can't sweep strategy"
-    amount: uint256 = 0
-    if token == ASSET.address:
-        amount = ASSET.balanceOf(self) - self.total_idle
-    else:
-        amount = ERC20(token).balanceOf(self)
-    assert amount != 0, "no dust"
-    self.erc20_safe_transfer(token, msg.sender, amount)
-    log Sweep(token, amount)
-    return amount
+def buy_debt(strategy: address, amount: uint256):
+    """
+    @notice Used for governance to buy bad debt from the vault.
+    @dev This should only ever be used in an emergency in place
+    of force revoking a strategy in order to not report a loss.
+    It allows the DEBT_PURCHASER role to buy the strategies debt
+    for an equal amount of `asset`. If the full amount of debt
+    is purchased revoke_strategy will be automatically called 
+    to remove the strategy.
+    @param strategy The strategy to buy the debt for
+    @param amount The amount of debt to buy from the vault.
+    """
+    self._enforce_role(msg.sender, Roles.DEBT_PURCHASER)
+    assert self.strategies[strategy].activation != 0, "not active"
+    
+    # cache the current debt
+    current_debt: uint256 = self.strategies[strategy].current_debt
+    
+    assert current_debt > 0, "nothing to buy"
+    assert amount > 0, "nothing to buy with"
+
+    # Get the current shares value for the amount
+    shares: uint256 = IStrategy(strategy).convertToShares(amount)
+    assert shares > 0, "can't buy 0"
+    assert shares <= IStrategy(strategy).balanceOf(self), "not enough shares"
+
+    before_balance: uint256 = ASSET.balanceOf(self)
+    self._erc20_safe_transfer_from(ASSET.address, msg.sender, self, amount)
+    after_balance: uint256 = ASSET.balanceOf(self)
+
+    assert after_balance - before_balance >= amount
+
+    # Adjust if needed to not underflow on math
+    bought: uint256 = min(current_debt, amount)
+
+    # Lower strategy debt
+    self.strategies[strategy].current_debt -= bought
+    # lower total debt
+    self.total_debt -= bought
+    # Increase total idle
+    self.total_idle += bought
+
+    # log debt change
+    log DebtUpdated(strategy, current_debt, current_debt - bought)
+
+    # If all the debt was bought revoke the strategy
+    if bought == current_debt:
+        self._revoke_strategy(strategy)
```

Comment:
Question:

This is a very logical design to revoke after debt is purchased.
But it does create a situation where DEBT_PURCHASER role can invoke actions explicitly designated for a different role. Maybe that's OK?

Reply:
your right it shouldnt do multiple jobs.